prog/weavedns/main.go

```diff
@@ -74,15 +77,44 @@ func main() {
 	var iface *net.Interface
 	if ifaceName != "" {
 		var err error
-		Info.Println("Waiting for interface", ifaceName, "to come up")
+		Info.Println("[main] Waiting for mDNS interface", ifaceName, "to come up")
 		iface, err = weavenet.EnsureInterface(ifaceName, wait)
 		if err != nil {
 			Error.Fatal(err)
 		} else {
-			Info.Println("Interface", ifaceName, "is up")
+			Info.Println("[main] Interface", ifaceName, "is up")
 		}
 	}
 
+	var httpAddr string
+	if httpIfaceName == "" {
+		httpAddr = net.JoinHostPort("0.0.0.0", strconv.Itoa(httpPort))
+	} else {
+		Info.Println("[main] Waiting for HTTP interface", httpIfaceName, "to come up")
+		httpIface, err := weavenet.EnsureInterface(httpIfaceName, wait)
+		if err != nil {
+			Error.Fatal(err)
+		}
+		Info.Println("[main] Interface", httpIfaceName, "is up")
+
+		addrs, err := httpIface.Addrs()
+		if err != nil {
+			Error.Fatal(err)
+		}
+
+		if len(addrs) == 0 {
+			Error.Fatal("[main] No addresses on HTTP interface")
+		}
+
+		ip, _, err := net.ParseCIDR(addrs[0].String())
+		if err != nil {
+			Error.Fatal(err)
+		}
+
+		httpAddr = net.JoinHostPort(ip.String(), strconv.Itoa(httpPort))
+	}
```

Comment:
The `if` should just compute an `httpIP` string; and the joining with the port should happen afterwards.